pageserver/src/tenant/timeline.rs

```diff
@@ -4759,31 +4705,177 @@ impl Deref for TimelineWriter<'_> {
+
+                let layer = self.tl.get_layer_for_write(at).await?;
+                let initial_size = layer.size().await?;
+                self.write_guard
+                    .replace(TimelineWriterState::new(layer, initial_size));
+            }
+            OpenLayerAction::Open => {
+                assert!(self.write_guard.is_none());
+
+                let layer = self.tl.get_layer_for_write(at).await?;
+                let initial_size = layer.size().await?;
+                self.write_guard
+                    .replace(TimelineWriterState::new(layer, initial_size));
+            }
+            OpenLayerAction::None => {
+                assert!(self.write_guard.is_some());
+            }
+        }
+
+        Ok(&self.write_guard.as_ref().unwrap().open_layer)
+    }
+
+    fn get_open_layer_action(&self, lsn: Lsn, new_value_size: u64) -> OpenLayerAction {
+        let state = &*self.write_guard;
+        if state.is_none() {
+            return OpenLayerAction::Open;
+        }
+
+        let state = state.as_ref().unwrap();
+        if state.prev_lsn == Some(lsn) {
+            // Rolling mid LSN is not supported by downstream code.
+            // Hence, only roll at LSN boundaries.
+            return OpenLayerAction::None;
+        }
+
+        let last_freeze_at = self.last_freeze_at.load();
+        let distance = lsn.widening_sub(last_freeze_at);
+        let last_freeze_ts = *(self.last_freeze_ts.read().unwrap());
+
+        let proposed_open_layer_size = state.current_size + new_value_size;
+
+        // Rolling the open layer can be triggered by:
+        // 1. The distance from the last LSN we rolled at. This bounds the amount of WAL that
+        //    the safekeepers need to store.
+        // 2. The size of the currently open layer.
+        // 3. The time since the last roll. It helps safekeepers to regard pageserver as caught
+        //    up and suspend activity.
+        if distance >= self.get_checkpoint_distance().into() {
+            info!(
+                "Will roll layer at {} with layer size {} due to LSN distance ({})",
+                lsn, state.current_size, distance
+            );
+
+            OpenLayerAction::Roll
+        } else if state.current_size > 0
+            && proposed_open_layer_size >= self.get_checkpoint_distance()
+        {
+            info!(
+                "Will roll layer at {} with layer size {} due to layer size ({})",
+                lsn, state.current_size, proposed_open_layer_size
+            );
+
+            OpenLayerAction::Roll
+        } else if distance > 0 && last_freeze_ts.elapsed() >= self.get_checkpoint_timeout() {
+            info!(
+                "Will roll layer at {} with layer size {} due to time since last flush ({:?})",
+                lsn,
+                state.current_size,
+                last_freeze_ts.elapsed()
+            );
+
+            OpenLayerAction::Roll
+        } else {
+            OpenLayerAction::None
+        }
```

Comment:
Would be nice to evaluate all these conditions, then determine the action, and then log a single  `info!()` variable inputs & decision. Because, this if else if else if cascade is really a logical OR, not something that _needs_ to be checked in that order.

Reply:
Right, but then the reason for the flush is not immediately apparent from the log line. I think I prefer the current approach, but don't feel strongly about it.

Reply:
I don't feel _super_ strongly either, but, the old message contained the distance, which helped my brain spot https://github.com/neondatabase/neon/issues/6631

https://github.com/neondatabase/neon/pull/6661/files#diff-74152cb7b3681c454c4156d530ce544ee8f15048c9f3cacc736dba5dc7c0d664L1056-L1062